### PR TITLE
[FLINK-21389] determine parquet schema from file instead of taking it from user

### PR DESCRIPTION
**Pull request by echauchot**
## What is the purpose of the change

[FLINK-21389] determine parquet schema from file instead of taking it from user

## Brief change log

deprecate ParquetInputFormat constructors that take schema as parameter and create a constructor without parquet schema as parameter. Determine the schema from the file in the constructor instead

## Verifying this change

This change added tests and can be verified as follows:
Added ParquetInputFormatTest for low level tests.
This change can be tested by ParquetInputFormatTest#testReadWithoutParquetSchemaSpecified()

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes deprecation
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes (avoid user having to pass parque schema to parquet input format)
  - If yes, how is the feature documented? deprecation in javadocs

### CODE REVIEW

**Comment by echauchot** [comment]
@HuangZhenQiu you already agreed for this change in the jira ticket. Still there is a question (in TODO in the code and in the ticket) why is this needed to read the parquet schema from split in `open()` ? Shouldn't it be the same schema than in the complete file ? Can't we use the already extracted schema in the constructor ?

@zentol as @HuangZhenQiu is not a committer, could you please take a look or point me to another committer ?

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
@zentol Hi, do you have time to review this or can you point me to another reviewer if not ?

**Comment by zentol** [comment]
maybe @JingsongLi could take a look; this is quite a bit out of my area of expertise

**Comment by echauchot** [comment]
@Arvid Heise can you also take a look at this PR ?

**Comment by AHeise** [comment]
Sure I have assigned myself, so I receive updates. But realistically, it will be next week ;).

**Comment by echauchot** [comment]
@AHeise same comment as in https://github.com/apache/flink/pull/15156#issuecomment-847073898 ParquetInputFormat.java was removed

**Comment by AHeise** [comment]
Same as in the other PR please rebase onto 1.13 and update the target branch accordingly.

A high-level question to get me faster up-to-speed:
- In Avro, we have reader and writer schema. If schema evolves, the writer schema of each record updates and through schema compability, I still get the equivalent record in the reader schema automatically. So for Avro, I'd usually specify an additional schema to make sure that my application is forward and backward compatible. 
- Now it seems like Parquet (haven't checked the details yet), there is a similar concept. Having a particular reader schema is even more important as it allows us to skip reading large chunks of the file if a specific column is not needed thanks to the columnar layout of the file.
- Is your change now effectively disabling the reader schema? Or can it just be omitted and assumed to be the writer schema?
- How would it work when I read 2 parquet files with different schemas but both can be mapped to the same reader schema? For example, consider a schema evolution case, where 1 file is written by pipeline v1 and file 2 is written by pipeline v2 with an additional column that is ignored in the consuming Flink application.

**Comment by echauchot** [comment]
rebased on release-1.13

**Comment by echauchot** [comment]
> Same as in the other PR please rebase onto 1.13 and update the target branch accordingly.
> 
> A high-level question to get me faster up-to-speed:
> 
> * In Avro, we have reader and writer schema. If schema evolves, the writer schema of each record updates and through schema compability, I still get the equivalent record in the reader schema automatically. So for Avro, I'd usually specify an additional schema to make sure that my application is forward and backward compatible.
> * Now it seems like Parquet (haven't checked the details yet), there is a similar concept. Having a particular reader schema is even more important as it allows us to skip reading large chunks of the file if a specific column is not needed thanks to the columnar layout of the file.
> * Is your change now effectively disabling the reader schema? Or can it just be omitted and assumed to be the writer schema?
> * How would it work when I read 2 parquet files with different schemas but both can be mapped to the same reader schema? For example, consider a schema evolution case, where 1 file is written by pipeline v1 and file 2 is written by pipeline v2 with an additional column that is ignored in the consuming Flink application.

My PR actually solves a bug: ParquetInputFormat takes parquet schema as user input but after split it reads the parquet schema again here https://github.com/apache/flink/blob/52dcf439bb0b8d613fff1efecf015052d5b3a10b/flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java#L170. So if a user provides a read schema then, he will be surprised that the actual schema used after split is different.

My PR touches only the reading part no change is done to the write schema.

Regarding avoiding read the file to determine the schema, extracting the schema from the file is actually already done in https://github.com/apache/flink/blob/52dcf439bb0b8d613fff1efecf015052d5b3a10b/flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java#L170. 


Why is this needed to read the parquet schema from Flink split in _open()_ ? Can't we use the already extracted schema (extracted in the constructor) ?

**Comment by echauchot** [comment]
@AHeise what do you think about the above comments?

**Comment by AHeise** [comment]
Sorry I'll check in the beginning of next week.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (AHeise) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:172]: See above.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (AHeise) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:181]: Here we read the write schema already. No need to add `readParquetSchemaFromFile`.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (AHeise) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:182]: Here we should check if `fieldNames` is null and fill them. If set, call `getReadSchema` as is.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetMapInputFormat.java
Comment (AHeise) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetMapInputFormat.java:48]: Remove comment, see high-level review comment.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetMapInputFormat.java
Comment (AHeise) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetMapInputFormat.java:50]: Since we do not check `messageType != null` anymore in base class, we need to add a `Preconditions#checkNotNull` here.

**Review by AHeise — CHANGES_REQUESTED** [review]
Hi @echauchot ,

sorry for the delay, I was busy guiding some FLIPs.

I have checked the PR but it's like I pointed out before:
- Your change is effectively removing reader schema. In the current state, you'd always receive reader=writer schema, which is not good in many cases.
- We can and should make reader schema optional. All new constructors are here to stay and ease using it, if you want to read all columns.
- Reading the writer schema currently happens in `#open` and it should stay like this. Quite often, the client which submits the application doesn't have access to the data it points to (think of GDPR compliance). So it's better to do it in `#open`, which is only called on task managers.
- Since the writer schema is already read in `#open`, you could populate the fields `fieldNames` and `fieldTypes` there in case of no explicit reader schema.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (echauchot) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:108]: "The constructor is not meant to do any I/O operations"
=> thank for the hint! Agreed

"I propose to simply leave the expectedFileSchema at null.
#open would then simply use reader schema = writer schema, which was probably you intention."
=> Fair enough. Yes that was my intention indeed

**Comment by echauchot** [comment]
> Hi @echauchot ,
> 
> sorry for the delay, I was busy guiding some FLIPs.
> 
> I have checked the PR but it's like I pointed out before:
> 
> * Your change is effectively removing reader schema. In the current state, you'd always receive reader=writer schema, which is not good in many cases.
> * We can and should make reader schema optional. All new constructors are here to stay and ease using it, if you want to read all columns.
> * Reading the writer schema currently happens in `#open` and it should stay like this. Quite often, the client which submits the application doesn't have access to the data it points to (think of GDPR compliance). So it's better to do it in `#open`, which is only called on task managers.
> * Since the writer schema is already read in `#open`, you could populate the fields `fieldNames` and `fieldTypes` there in case of no explicit reader schema.

Thanks for the precision about reader schema and writer schema. That makes sense to allow to read using a different schema that the one that was used when writing the file. But still, there is a thing I don't get: what is the point of providing a schema in _ParquetInputFormat_ constructor, it is not the one that will be used **at the actual reading time**, it is the writer schema (the one extracted in _#open_) that will be used. Hence the fact that I qualified the ticket as a bug with current flink master state.

I agree IO operations should only happen en task managers as the job manager might not have access to data (network filtering or other) and for parallelization also I guess.

When no schema is provided,  of course, we can set the fields with values read in _#open_

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (echauchot) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:172]: Yes, clearer now, thanks. But still I think there is a bug, see my comment at the end of the review.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (echauchot) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:181]: yes, I got it, only read the schema once in _#open_

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetMapInputFormat.java
Comment (echauchot) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetMapInputFormat.java:48]: yes

**Comment by AHeise** [comment]
>  But still, there is a thing I don't get: what is the point of providing a schema in _ParquetInputFormat_ constructor, it is not the one that will be used **at the actual reading time**, it is the writer schema (the one extracted in _#open_) that will be used. Hence the fact that I qualified the ticket as a bug with current flink master state.

Have a look at `ParquetTableSource` and the used `ParquetRowInputFormat`. The idea is that the projection is pushed into the source. So if you have a `SELECT a,b FROM parquet_file`, you really just need to read `a, b`. In columnar formats, that translates to omitting a large chunk of data.

In the other formats, a user would need to supply the desired read schema manually. That is probably quite rare, but I'd keep it as it's just a small constructor each.

**Comment by echauchot** [comment]
@AHeise I addressed your comments PTAL.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (AHeise) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:100]: Please annotate the nullable fields no with `@Nullable` and add a comment like, `can be null if it should be inferred from the file` or so.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (AHeise) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:188]: :+1: you even covered the hard case with `selectFields`.

**Review by AHeise — COMMENTED** [review]
Thanks again. LGTM. I'll approve as soon as the build turns green.
Please squash the 3 commits into one before the merge (I can also do that while merging).

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (echauchot) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:100]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java
Comment (echauchot) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java:188]: yes :)

**Comment by echauchot** [comment]
Addressed your comments and squashed. PTAL

**Review by AHeise — APPROVED** [review]
LGTM! Merging.

**Comment by echauchot** [comment]
@AHeise thanks a lot for reviewing and merging !

**Comment by AHeise** [comment]
Thank you for contributing!

**Pull request merged**
